include/deal.II/differentiation/ad/ad_drivers.h

```diff
@@ -0,0 +1,1094 @@
+      const bool &)
+    {
+      AssertThrow(false, ExcNotImplemented());
+    }
+
+
+    template<typename ADNumberType, typename ScalarType, typename T>
+    ScalarType
+    TapedDrivers<ADNumberType,ScalarType,T>::value (
+      const unsigned int &,
+      const unsigned int &,
+      const std::vector<ScalarType> &)
+    {
+      AssertThrow(false, ExcNotImplemented());
+      return ScalarType(0.0);
+    }
+
+
+    template<typename ADNumberType, typename ScalarType, typename T>
+    void
+    TapedDrivers<ADNumberType,ScalarType,T>::gradient (
+      const unsigned int &,
+      const unsigned int &,
+      const std::vector<ScalarType> &,
+      Vector<ScalarType> &)
+    {
+      AssertThrow(false, ExcNotImplemented());
+    }
+
+
+    template<typename ADNumberType, typename ScalarType, typename T>
+    void
+    TapedDrivers<ADNumberType,ScalarType,T>::hessian (
+      const unsigned int &,
+      const unsigned int &,
+      const std::vector<ScalarType> &,
+      FullMatrix<ScalarType> &)
+    {
+      AssertThrow(false, ExcNotImplemented());
+    }
+
+
+    template<typename ADNumberType, typename ScalarType, typename T>
+    void
+    TapedDrivers<ADNumberType,ScalarType,T>::values (
+      const unsigned int &,
+      const unsigned int &,
+      const unsigned int &,
+      const std::vector<ScalarType> &,
+      Vector<ScalarType> &)
+    {
+      AssertThrow(false, ExcNotImplemented());
+    }
+
+
+    template<typename ADNumberType, typename ScalarType, typename T>
+    void
+    TapedDrivers<ADNumberType,ScalarType,T>::jacobian (
+      const unsigned int &,
+      const unsigned int &,
+      const unsigned int &,
+      const std::vector<ScalarType> &,
+      FullMatrix<ScalarType> &)
+    {
+      AssertThrow(false, ExcNotImplemented());
+    }
+
+
+    // Specialization for taped Adol-C auto-differentiable numbers.
+    //
+    // Note: In the case of Adol-C taped numbers, the associated scalar
+    // type is always expected to be a double. So we need to make a further
+    // specialization when ScalarType is a float.
+    template<typename ADNumberType>
+    struct TapedDrivers<ADNumberType,double,typename std::enable_if<
+      ADNumberTraits<ADNumberType>::type_code == NumberTypes::adolc_taped
+      >::type>
+    {
+      typedef double scalar_type;
+
+      // === Taping ===
+
+      static void
+      enable_taping(const unsigned int &tape_index,
+                    const bool         &keep)
+      {
+        trace_on(tape_index,keep);
+      }
+
+      static void
+      disable_taping(const unsigned int &active_tape_index,
+                     const bool         &write_tapes_to_file)
+      {
+        if (write_tapes_to_file)
+          {
+            trace_off(active_tape_index); // Slow
+
+            std::vector<std::size_t> counts (STAT_SIZE);
+#ifdef DEAL_II_WITH_ADOLC
+            ::tapestats(active_tape_index, counts.data());
+#else
+            AssertThrow(false, ExcNotImplemented());
```

Comment:
here and below maybe introduce a custom exception with the message along the lines "This function is only available if deal.II is compiled with ADOL-C".

Reply:
As per your suggestion, I've added a custom error message.